File: layoutz-hs/CounterDemo.hs
{-# LANGUAGE OverloadedStrings #-}

{-|
Simple counter demo app using Layoutz TUI runtime.

Press '+' to increment, '-' to decrement.
Press ESC, Ctrl+C, or Ctrl+D to quit.
-}

module Main where

import Layoutz

-- | Messages that can change counter state
data CounterMsg = Inc | Dec

-- | The counter application
counterApp :: LayoutzApp Int CounterMsg
counterApp = LayoutzApp
  { appInit = (0, CmdNone)
  
  , appUpdate = \msg count -> case msg of
      Inc -> (count + 1, CmdNone)
      Dec -> (count - 1, CmdNone)
  
  , appSubscriptions = \_state ->
      subKeyPress $ \key -> case key of
        KeyChar '+' -> Just Inc
        KeyChar '-' -> Just Dec
        _           -> Nothing
  
  , appView = \count ->
      layout
        [ section "Counter" [text $ "Count: " <> show count]
        , br
        , ul ["Press '+' or '-'", "Press ESC to quit"]
        ]
  }

main :: IO ()
main = runApp counterApp

File: layoutz-ocaml/counter_demo.ml
open Layoutz

(* A simple counter app demonstrating the Elm Architecture runtime *)

type msg = Inc | Dec | Tick

let () =
  run_app
    {
      init = (0, CmdNone);
      update =
        (fun msg count ->
          match msg with
          | Inc -> (count + 1, CmdNone)
          | Dec -> (count - 1, CmdNone)
          | Tick -> (count, CmdNone));
      subscriptions =
        (fun _ ->
          sub_batch
            [
              sub_key_press (fun key ->
                  match key with
                  | KeyUp -> Some Inc
                  | KeyDown -> Some Dec
                  | KeyChar '+' -> Some Inc
                  | KeyChar '-' -> Some Dec
                  | _ -> None);
              sub_every_ms 1000 Tick;
            ]);
      view =
        (fun count ->
          layout
            [
              center (s "Counter Demo" |> styleBold |> colorBrightCyan);
              s "";
              row
                [
                  statusCard ~label:(s "Count")
                    ~content:(s (string_of_int count))
                  |> borderRound
                  |> colorBrightGreen;
                ];
              s "";
              s "  Up/+ to increment, Down/- to decrement" |> colorBrightBlack;
              s "  Ctrl+Q to quit" |> colorBrightBlack;
            ]);
    }
